File: backend/app/models/buddies.py
"""
FitTracker Backend - 搭子相关数据模型
"""

from sqlalchemy import Column, Integer, String, Text, Boolean, DateTime, ForeignKey, JSON, Float
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from app.core.database import Base

class WorkoutBuddy(Base):
    """健身搭子关系表"""
    __tablename__ = "workout_buddies"

    id = Column(Integer, primary_key=True, index=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
    
    # 用户关系
    user_id = Column(String, ForeignKey("users.id"), nullable=False)
    buddy_id = Column(String, ForeignKey("users.id"), nullable=False)
    
    # 关系状态
    status = Column(String, default="active")  # active, paused, ended
    
    # 匹配信息
    match_score = Column(Integer, default=0)  # 匹配度分数 0-100
    match_reasons = Column(JSON, default=list)  # 匹配原因列表
    
    # 训练偏好
    workout_preferences = Column(JSON, default=dict)  # 训练偏好
    location_match = Column(Boolean, default=False)  # 地理位置匹配
    schedule_match = Column(Boolean, default=False)  # 时间安排匹配
    goal_match = Column(Boolean, default=False)  # 健身目标匹配
    
    # 关系管理
    last_interaction = Column(DateTime(timezone=True))
    total_workouts = Column(Integer, default=0)  # 一起训练次数
    rating = Column(Float, default=0.0)  # 搭子评分
    
    # 约束
    __table_args__ = (
        {"extend_existing": True}
    )

class BuddyRequest(Base):
    """搭子申请表"""
    __tablename__ = "buddy_requests"

    id = Column(Integer, primary_key=True, index=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
    
    # 申请关系
    requester_id = Column(String, ForeignKey("users.id"), nullable=False)
    target_id = Column(String, ForeignKey("users.id"), nullable=False)
    
    # 申请状态
    status = Column(String, default="pending")  # pending, accepted, rejected, expired
    
    # 申请信息
    request_message = Column(Text)
    response_message = Column(Text)
    
    # 训练偏好
    workout_preferences = Column(JSON, default=dict)
    preferred_time = Column(String)  # 偏好训练时间
    preferred_location = Column(String)  # 偏好训练地点
    
    # 时间管理
    requested_at = Column(DateTime(timezone=True), server_default=func.now())
    responded_at = Column(DateTime(timezone=True))
    expires_at = Column(DateTime(timezone=True))  # 申请过期时间
    
    # 约束
    __table_args__ = (
        {"extend_existing": True}
    )

class BuddyMatch(Base):
    """搭子匹配记录表"""
    __tablename__ = "buddy_matches"

    id = Column(Integer, primary_key=True, index=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    
    # 匹配关系
    user_id = Column(String, ForeignKey("users.id"), nullable=False)
    matched_user_id = Column(String, ForeignKey("users.id"), nullable=False)
    
    # 匹配信息
    match_score = Column(Integer, nullable=False)  # 匹配分数
    match_reasons = Column(JSON, default=list)  # 匹配原因
    match_type = Column(String)  # 匹配类型: recommendation, nearby, similar
    
    # 匹配状态
    is_viewed = Column(Boolean, default=False)  # 是否已查看
    is_interested = Column(Boolean, default=False)  # 是否感兴趣
    
    # 约束
    __table_args__ = (
        {"extend_existing": True}
    )

File: backend-go/internal/models/buddies.go
package models

import (
	"time"
)

// BuddyRequest 搭子申请模型
type BuddyRequest struct {
	ID                 uint      `json:"id" gorm:"primaryKey,autoIncrement"`
	RequesterID        string    `json:"requester_id" gorm:"not null"`
	ReceiverID         string    `json:"receiver_id" gorm:"not null"`
	Message            string    `json:"message" gorm:"type:text"`
	Status             string    `json:"status" gorm:"default:'pending'"` // pending, accepted, rejected
	WorkoutPreferences string    `json:"workout_preferences" gorm:"type:json"`
	CreatedAt          time.Time `json:"created_at"`
	UpdatedAt          time.Time `json:"updated_at"`

	// 关联关系
	Requester User `json:"requester" gorm:"foreignKey:RequesterID"`
	Receiver  User `json:"receiver" gorm:"foreignKey:ReceiverID"`
}

// BuddyRelationship 搭子关系模型
type BuddyRelationship struct {
	ID            uint      `json:"id" gorm:"primaryKey,autoIncrement"`
	CreatedAt     time.Time `json:"created_at"`
	UpdatedAt     time.Time `json:"updated_at"`
	UserID        string    `json:"user_id" gorm:"not null"`
	BuddyID       string    `json:"buddy_id" gorm:"not null"`
	EstablishedAt time.Time `json:"established_at"`
	Status        string    `json:"status" gorm:"default:'active'"` // active, blocked

	// 关联关系
	User  User `json:"user" gorm:"foreignKey:UserID"`
	Buddy User `json:"buddy" gorm:"foreignKey:BuddyID"`
}

// WorkoutPreferences 训练偏好
type WorkoutPreferences struct {
	Time     string `json:"time"`     // 训练时间偏好
	Location string `json:"location"` // 训练地点偏好
	Type     string `json:"type"`     // 训练类型偏好
}

// BuddyRequestCreate 创建搭子申请请求
type BuddyRequestCreate struct {
	ReceiverID         string              `json:"receiver_id" binding:"required"`
	Message            string              `json:"message"`
	WorkoutPreferences *WorkoutPreferences `json:"workout_preferences"`
}

// BuddyRequestResponse 搭子申请响应
type BuddyRequestResponse struct {
	ID                 uint                `json:"id"`
	RequesterID        string              `json:"requester_id"`
	ReceiverID         string              `json:"receiver_id"`
	Message            string              `json:"message"`
	Status             string              `json:"status"`
	WorkoutPreferences *WorkoutPreferences `json:"workout_preferences"`
	CreatedAt          time.Time           `json:"created_at"`
	UpdatedAt          time.Time           `json:"updated_at"`
	Requester          *User               `json:"requester"`
	Receiver           *User               `json:"receiver"`
}

// BuddyRecommendationResponse 搭子推荐响应
type BuddyRecommendationResponse struct {
	User               User                `json:"user"`
	MatchScore         int                 `json:"match_score"`
	MatchReasons       []string            `json:"match_reasons"`
	WorkoutPreferences *WorkoutPreferences `json:"workout_preferences"`
}

// BuddyResponse 搭子关系响应
type BuddyResponse struct {
	ID            uint      `json:"id"`
	UserID        string    `json:"user_id"`
	BuddyID       string    `json:"buddy_id"`
	EstablishedAt time.Time `json:"established_at"`
	Status        string    `json:"status"`
	User          *User     `json:"user"`
	Buddy         *User     `json:"buddy"`
}

// BuddyUpdateRequest 更新搭子关系请求
type BuddyUpdateRequest struct {
	Status  string `json:"status"` // accepted, rejected, blocked
	Message string `json:"message"`
}
